sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -874,6 +874,100 @@ CURLcode CurlSession::ReadStatusLineAndHeadersFromRawResponse(
   this->m_innerBufferSize = bufferSize;
   this->m_lastStatusCode = this->m_response->GetStatusCode();
 
+  // The logic below comes from the expectation that Azure services, particularly Storage, may not
+  // conform to HTTP standards when it comes to handling 100-continue requests, and not send
+  // "Connection: close" when they should. We do not know for sure if this is true, but this logic
+  // did exist for libcurl transport in earlier C++ SDK versions.
+  //
+  // The idea is the following: if status code is not 2xx, and request header contains "Expect:
+  // 100-continue" and request body length is not zero, we don't reuse the connection.
+  //
+  // More detailed description of what might happen if we don't have this logic:
+  // 1. Storage SDK sends a PUT request with a non-empty request body (which means Content-Length
+  // request header is not 0, let's say it's 6) and Expect: 100-continue request header, but it
+  // doesn't send the header unless server returns 100 Continue status code.
+  // 2. Storage service returns 4xx status code and response headers, but it doesn't want to close
+  // this connection, so there's no Connection: close in response headers.
+  // 3. Now both client and server agree to continue using this connection. But they do not agree in
+  // the current status of this connection.
+  //    3.1. Client side thinks the previous request is finished because it has received a status
+  //    code and response headers. It should send a new HTTP request if there's any.
+  //    3.2. Server side thinks the previous request is not finished because it hasn't received the
+  //    request body. I tend to think this is a bug of server-side.
+  // 4. Client side sends a new request, for example,
+  //    HEAD /whatever/path HTTP/1.1
+  //    host: foo.bar.com
+  //    ...
+  // 5. Server side takes the first 6 bytes (HEAD /) of the send request and thinks this is the
+  // request body of the first request and discard it.
+  // 6. Server side keeps reading the remaining data on the wire and thinks the first part
+  // (whatever/path) is an HTTP verb. It fails the request with 400 invalid verb.
+  bool non2xxAfter100ContinueWithNonzeroContentLength = false;
+  {
```

Comment:
Thanks! This is good feedback. Let me to address it in the follow-up PR that I will make shortly, so we could potentially debate the better place for the code, while this logic, that I got the green light for, can be checked in.
I do agree, the place you suggesting seems like more natural place to detect the condition, the only difference being that if "100-continue" is set in some other place in code, and the verb is not "PUT", then it may not detect it, but from what I see, we are not doing that anywhere in our code, other than that place inside libcurl transport adapter.

Reply:
@mchelnokov, I made another PR - https://github.com/antkmsft/azure-sdk-for-cpp/pull/1. Would you say it reflects the idea you had in mind?

@Jinming-Hu, would you prefer the logic from this PR, or from https://github.com/antkmsft/azure-sdk-for-cpp/pull/1?
I realize that, for better, or for worse, it is not a 100% equivalent condition-wise.